go/7_auth-azure.go

```diff
@@ -0,0 +1,184 @@
+			"resource_group_name" : metadataRespJson.Compute.ResourceGroupName, 
+	}
+
+	// log in to Vault's Azure auth method
+	resp, err := client.Logical().Write("auth/azure/login", params) // confirm with your Vault administrator that "azure" is the correct mount name
+	if err != nil {
+		return "", fmt.Errorf("unable to log in with Azure auth: %w", err)
+	}
+	if resp == nil || resp.Auth == nil || resp.Auth.ClientToken == "" {
+		return "", fmt.Errorf("login response did not return client token")
+	}
+
+	client.SetToken(resp.Auth.ClientToken)
+	// get secret
+	secret, err := client.Logical().Read("kv-v2/data/creds")
+	if err != nil {
+		return "", fmt.Errorf("unable to read secret: %w", err)
+	}
+
+	data, ok := secret.Data["data"].(map[string]interface{})
+	if !ok {
+		return "", fmt.Errorf("data type assertion failed: %T %#v", secret.Data["data"], secret.Data["data"])
+	}
+
+	// data map can contain more than one key-value pair, in this case we're just grabbing one of them
+	key := "password"
+	value, ok := data[key].(string)
+	if !ok {
+		return "", fmt.Errorf("value type assertion failed: %T %#v", data[key], data[key])
+	}
+
+	return value, nil
+}
+
+// Retrieve instance metadata from Azure 
+func getMetadata() (metadataJson, error) {
+
+	var metadata_endpoint *url.URL
+	metadata_endpoint, err := url.Parse("http://169.254.169.254/metadata/instance?api-version=2017-08-01")
+	if err != nil {
+  		fmt.Println("Error creating URL: ", err)
+  		return metadataJson{}, err
+	}
+
+	metadata_parameters := metadata_endpoint.Query()
+	metadata_endpoint.RawQuery = metadata_parameters.Encode()
+	req, err := http.NewRequest("GET", metadata_endpoint.String(), nil)
+	if err != nil {
+		return metadataJson{}, fmt.Errorf("Error creating HTTP Request: %w", err)
+	}
+	req.Header.Add("Metadata", "true")
+
+    client := &http.Client{}
+    resp, err := client.Do(req) 
+    if err != nil{
+      fmt.Println("Error calling token endpoint: ", err)
+      return metadataJson{}, fmt.Errorf("Error calling token endpoint: %w", err)
+    }
+
+    responseBytes,err := ioutil.ReadAll(resp.Body)
+    defer resp.Body.Close()
+    if err != nil {
+      return metadataJson{}, fmt.Errorf("Error reading response body: %w", err)
+    }
+
+	// Unmarshal response body into metadata struct 
+	var r metadataJson
+    err = json.Unmarshal(responseBytes, &r)
+    if err != nil {
+      return metadataJson{}, fmt.Errorf("Error unmarshalling the response: %w", err)
+    }
+
+	return r, nil	
+}
+
+// Retrieves an access token from Azure MSI
+func getJWT() (string, error) {
+	// Create HTTP request for a managed services for Azure resources token to access Azure Resource Manager
+    var msi_endpoint *url.URL
+    msi_endpoint, err := url.Parse("http://169.254.169.254/metadata/identity/oauth2/token?api-version=2018-02-01")
+    if err != nil {
+      return "", fmt.Errorf("Error creating URL: %w", err)
+
+    }
+
+	msi_parameters := msi_endpoint.Query()
```

Comment:
Another extra tab here. Do you have the Go extension for VSCode installed? I feel like it should be fixing indentation issues on save